### ISSUE TITLE
Blocking subscriber.onComplete called after subscriber.onError

### ISSUE THREAD

**Issue by vemilyus** [jOOQ/jOOQ#10245]
### Expected behavior and actual behavior:
The expected behavior is to call either `onComplete` or `onError`, never both, one after the other.

Actually it always calls `onComplete`, even if `onError` was called before. 

See here for one particular offender I found: https://github.com/jOOQ/jOOQ/blob/934d96f3b1413169ebe5e88cc4f7c9eac8846a04/jOOQ/src/main/java/org/jooq/impl/AbstractResultQuery.java#L384-L396

A solution for this would be to make the `onComplete` call in `doComplete` conditional to better fit the reactive-streams specification. 

I know both `onComplete` and `onError` are terminal operations and that subscribers should ignore everything that comes after either one, but to remove this one `onComplete` call would add some tolerance for subscribers that don't properly follow the reactive-streams specification.

### Versions:

- jOOQ: 3.13.2
- Java: 11
- Database (include vendor): PostgreSQL
- OS: WSL
- JDBC Driver (include name if inofficial driver): postgresql 42.2.12 (jdbc 4.2)

**Comment by vemilyus** [comment]
Mentioned here: https://github.com/Kotlin/kotlinx.coroutines/issues/2079

**Comment by lukaseder** [comment]
Thanks a lot for the report. I'll look into this shortly.

**Comment by bsideup** [comment]
FYI the Reactive Streams TCK should be able to detect such issues. If you have custom `Publisher`s, consider running the TCK as a low hanging option to verify it.

**Comment by lukaseder** [comment]
I've fixed this now in 3.15.0 and 3.14.9 (#11742) and will soon look into using the TCK to validate the implementation.

Note that in jOOQ 3.15.0, there will be a better, non-blocking solution based on R2DBC, which is also going to be verified by the TCK: https://github.com/jOOQ/jOOQ/issues/11700

**Comment by lukaseder** [comment]
Haha, yes, the TCK is revealing 😁. Let's get this right first, before backporting.

**Comment by bsideup** [comment]
@lukaseder :D FYI TCK does not check for blocking calls, but adding http://github.com/reactor/BlockHound would shine some light :)

**Comment by lukaseder** [comment]
Well, this particular issue here is about the jOOQ 3.14 blocking implementation (JDBC backed), so blocking is kinda acceptable, I think? But it would be a good idea to use such tools for the new R2DBC backed implementation.

**Comment by lukaseder** [comment]
Another issue that the TCK revealed, we're buffering way too much, and aren't really streaming:

```java
                if (buffer.size() < i)
                    buffer.addAll(c.fetchNext(i - buffer.size()));

                boolean complete = buffer.size() < i;
                while (!buffer.isEmpty())
                    subscriber.onNext(buffer.pollFirst());
```

The `required_spec317_mustNotSignalOnErrorWhenPendingAboveLongMaxValue()` test put quite some stress on my computer 😀

**Comment by lukaseder** [comment]
OK, great. I have confidence in the fix now and will backport to 3.14.9, releasing today. Any feedback by @vemilyus on the original kotlin coroutines use-case would be highly appreciated!
